File: scripts/probe-popcornmovies.py
"""
One-off probe for popcornmovies.io (CF challenge + player network map).
Usage: python scripts/probe-popcornmovies.py
"""
from __future__ import annotations

import json
import re
import sys
from pathlib import Path

from playwright.sync_api import sync_playwright

OUT = Path(__file__).resolve().parents[1] / "tmp" / "popcornmovies-probe.json"
HOME = "https://popcornmovies.io/"
CANDIDATES = [
    "https://popcornmovies.io/movie/27205",
    "https://popcornmovies.io/movies/27205",
    "https://popcornmovies.io/watch/movie/27205",
    "https://popcornmovies.io/watch/27205",
    "https://popcornmovies.io/film/27205",
    "https://popcornmovies.io/title/27205",
    "https://popcornmovies.io/movie/inception",
    "https://popcornmovies.io/movies/inception-2010",
]


def interesting(url: str) -> bool:
    u = url.lower()
    keys = (
        "m3u8",
        "mp4",
        "playlist",
        "embed",
        "stream",
        "vidsrc",
        "2embed",
        "proxy",
        "/api/",
        "tmdb",
        "subtitle",
        "vtt",
        "player",
        "source",
        "hls",
    )
    return any(k in u for k in keys)


def main() -> int:
    OUT.parent.mkdir(parents=True, exist_ok=True)
    result: dict = {
        "home": {},
        "routes": [],
        "network": [],
        "iframes": [],
        "page_links": [],
        "notes": [],
    }

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(
            user_agent=(
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                "AppleWebKit/537.36 (KHTML, like Gecko) "
                "Chrome/122.0.0.0 Safari/537.36"
            ),
            viewport={"width": 1365, "height": 900},
            locale="en-US",
        )
        page = context.new_page()

        def on_response(resp):
            try:
                url = resp.url
                if not interesting(url) and resp.request.resource_type not in (
                    "xhr",
                    "fetch",
                    "document",
                    "media",
                ):
                    return
                entry = {
                    "status": resp.status,
                    "type": resp.request.resource_type,
                    "method": resp.request.method,
                    "url": url[:500],
                    "ctype": (resp.headers or {}).get("content-type", "")[:80],
                }
                if interesting(url) or resp.request.resource_type in ("xhr", "fetch"):
                    result["network"].append(entry)
            except Exception as exc:  # noqa: BLE001
                result["notes"].append(f"response hook error: {exc}")

        page.on("response", on_response)

        print(">> home")
        page.goto(HOME, wait_until="domcontentloaded", timeout=90_000)
        page.wait_for_timeout(8_000)
        title = page.title()
        html = page.content()
        challenged = "Just a moment" in title or "cf-browser-verification" in html
        result["home"] = {
            "url": page.url,
            "title": title,
            "challenged": challenged,
            "html_len": len(html),
            "snippet": re.sub(r"\s+", " ", html)[:400],
        }
        print(f"   title={title!r} challenged={challenged} url={page.url}")

        if challenged:
            print(">> waiting for CF challenge...")
            try:
                page.wait_for_function(
                    "() => !document.title.includes('Just a moment')",
                    timeout=45_000,
                )
                page.wait_for_timeout(3_000)
                result["home"]["after_challenge"] = {
                    "title": page.title(),
                    "url": page.url,
                }
                print(f"   cleared -> {page.title()!r}")
            except Exception as exc:  # noqa: BLE001
                result["notes"].append(f"CF wait failed: {exc}")
                print(f"   CF wait failed: {exc}")

        # Collect homepage anchors for route discovery
        try:
            hrefs = page.eval_on_selector_all(
                "a[href]",
                "els => els.map(e => e.getAttribute('href')).filter(Boolean).slice(0, 80)",
            )
            result["page_links"] = hrefs
            print(f"   links={len(hrefs)}")
        except Exception as exc:  # noqa: BLE001
            result["notes"].append(f"link scrape failed: {exc}")

        # Try candidate watch/detail routes
        for url in CANDIDATES:
            print(f">> try {url}")
            try:
                resp = page.goto(url, wait_until="domcontentloaded", timeout=45_000)
                page.wait_for_timeout(4_000)
                frames = [f.url for f in page.frames if f.url and f.url != page.url]
                entry = {
                    "url": url,
                    "final": page.url,
                    "status": resp.status if resp else None,
                    "title": page.title(),
                    "iframes": frames[:20],
                    "challenged": "Just a moment" in page.title(),
                }
                result["routes"].append(entry)
                result["iframes"].extend(frames)
                print(f"   status={entry['status']} title={entry['title']!r} frames={len(frames)}")
                # If we landed on a real page, poke play buttons
                if not entry["challenged"] and entry["status"] and entry["status"] < 400:
                    for sel in [
                        "button:has-text('Play')",
                        "button:has-text('Watch')",
                        "[class*='play']",
                        "video",
                    ]:
                        try:
                            loc = page.locator(sel).first
                            if loc.count() and loc.is_visible():
                                loc.click(timeout=2000)
                                page.wait_for_timeout(3000)
                                break
                        except Exception:
                            pass
                    break
            except Exception as exc:  # noqa: BLE001
                result["routes"].append({"url": url, "error": str(exc)})
                print(f"   error: {exc}")

        # Dedupe network
        seen = set()
        uniq = []
        for n in result["network"]:
            key = (n.get("method"), n.get("status"), n.get("url"))
            if key in seen:
                continue
            seen.add(key)
            uniq.append(n)
        result["network"] = uniq[:200]
        result["iframes"] = list(dict.fromkeys(result["iframes"]))[:50]

        browser.close()

    OUT.write_text(json.dumps(result, indent=2), encoding="utf-8")
    print(f"\nWrote {OUT}")
    print(f"network hits={len(result['network'])} routes={len(result['routes'])}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/probe-popcornmovies.mjs
/**
 * Probe popcornmovies.io behind Cloudflare using Playwright.
 * Run: npx --yes -p playwright@1.49.1 node scripts/probe-popcornmovies.mjs
 */
import { chromium } from 'playwright'
import { mkdir, writeFile } from 'node:fs/promises'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'

const ROOT = join(dirname(fileURLToPath(import.meta.url)), '..')
const OUT = join(ROOT, 'tmp', 'popcornmovies-probe.json')

const CANDIDATES = [
  'https://popcornmovies.io/movie/27205',
  'https://popcornmovies.io/movies/27205',
  'https://popcornmovies.io/watch/movie/27205',
  'https://popcornmovies.io/watch/27205',
  'https://popcornmovies.io/film/27205',
  'https://popcornmovies.io/title/27205',
  'https://popcornmovies.io/movie/inception',
  'https://popcornmovies.io/movies/inception-2010',
]

function interesting(url) {
  const u = url.toLowerCase()
  return [
    'm3u8',
    'mp4',
    'playlist',
    'embed',
    'stream',
    'vidsrc',
    '2embed',
    'proxy',
    '/api/',
    'tmdb',
    'subtitle',
    'vtt',
    'player',
    'source',
    'hls',
  ].some((k) => u.includes(k))
}

const result = {
  home: {},
  routes: [],
  network: [],
  iframes: [],
  page_links: [],
  notes: [],
}

const browser = await chromium.launch({ headless: true })
const context = await browser.newContext({
  userAgent:
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
  viewport: { width: 1365, height: 900 },
  locale: 'en-US',
})
const page = await context.newPage()

page.on('response', async (resp) => {
  try {
    const url = resp.url()
    const type = resp.request().resourceType()
    if (!interesting(url) && !['xhr', 'fetch', 'document', 'media'].includes(type)) return
    if (interesting(url) || type === 'xhr' || type === 'fetch' || type === 'media') {
      result.network.push({
        status: resp.status(),
        type,
        method: resp.request().method(),
        url: url.slice(0, 500),
        ctype: (resp.headers()['content-type'] || '').slice(0, 80),
      })
    }
  } catch (e) {
    result.notes.push(`response hook: ${e.message}`)
  }
})

console.log('>> home')
await page.goto('https://popcornmovies.io/', { waitUntil: 'domcontentloaded', timeout: 90_000 })
await page.waitForTimeout(8000)
let title = await page.title()
let html = await page.content()
let challenged = title.includes('Just a moment') || html.includes('challenge-platform')
result.home = {
  url: page.url(),
  title,
  challenged,
  html_len: html.length,
  snippet: html.replace(/\s+/g, ' ').slice(0, 400),
}
console.log(`   title=${title} challenged=${challenged}`)

if (challenged) {
  console.log('>> waiting for CF...')
  try {
    await page.waitForFunction(() => !document.title.includes('Just a moment'), {
      timeout: 60_000,
    })
    await page.waitForTimeout(4000)
    result.home.after_challenge = { title: await page.title(), url: page.url() }
    console.log(`   cleared -> ${result.home.after_challenge.title}`)
  } catch (e) {
    result.notes.push(`CF wait failed: ${e.message}`)
    console.log(`   CF wait failed: ${e.message}`)
  }
}

try {
  result.page_links = await page.$$eval('a[href]', (els) =>
    els.map((e) => e.getAttribute('href')).filter(Boolean).slice(0, 100),
  )
  console.log(`   links=${result.page_links.length}`)
} catch (e) {
  result.notes.push(`links: ${e.message}`)
}

for (const url of CANDIDATES) {
  console.log(`>> try ${url}`)
  try {
    const resp = await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 45_000 })
    await page.waitForTimeout(4000)
    const frames = page.frames().map((f) => f.url()).filter((u) => u && u !== page.url())
    const entry = {
      url,
      final: page.url(),
      status: resp?.status() ?? null,
      title: await page.title(),
      iframes: frames.slice(0, 20),
      challenged: (await page.title()).includes('Just a moment'),
    }
    result.routes.push(entry)
    result.iframes.push(...frames)
    console.log(`   status=${entry.status} title=${entry.title} frames=${frames.length}`)

    if (!entry.challenged && entry.status && entry.status < 400) {
      for (const sel of ["button:has-text('Play')", "button:has-text('Watch')", '[class*="play"]']) {
        try {
          const loc = page.locator(sel).first
          if ((await loc.count()) > 0 && (await loc.isVisible())) {
            await loc.click({ timeout: 2000 })
            await page.waitForTimeout(4000)
            break
          }
        } catch {
          /* ignore */
        }
      }
      break
    }
  } catch (e) {
    result.routes.push({ url, error: e.message })
    console.log(`   error: ${e.message}`)
  }
}

const seen = new Set()
result.network = result.network
  .filter((n) => {
    const key = `${n.method}|${n.status}|${n.url}`
    if (seen.has(key)) return false
    seen.add(key)
    return true
  })
  .slice(0, 200)
result.iframes = [...new Set(result.iframes)].slice(0, 50)

await browser.close()
await mkdir(dirname(OUT), { recursive: true })
await writeFile(OUT, JSON.stringify(result, null, 2))
console.log(`\nWrote ${OUT}`)
console.log(`network=${result.network.length} routes=${result.routes.length}`)
